File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/event/WorldEventTest.kt
package org.wyrdsekai.app.engine.event

import kotlin.time.Clock
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class WorldEventTest {

    private val now = Clock.System.now()

    @Test
    fun roomCreatedConstruction() {
        val event = WorldEvent.RoomCreated(
            roomId = "room-1",
            timestamp = now,
            name = "The Nexus",
            description = "A vast crystalline chamber.",
            zone = "foundation",
        )
        assertEquals("room-1", event.roomId)
        assertEquals(now, event.timestamp)
        assertEquals("The Nexus", event.name)
        assertEquals("A vast crystalline chamber.", event.description)
        assertEquals("foundation", event.zone)
    }

    @Test
    fun saidConstruction() {
        val event = WorldEvent.Said(
            roomId = "room-1",
            timestamp = now,
            entityId = "player-1",
            entityName = "Alice",
            text = "Hello, world!",
        )
        assertEquals("room-1", event.roomId)
        assertEquals(now, event.timestamp)
        assertEquals("player-1", event.entityId)
        assertEquals("Alice", event.entityName)
        assertEquals("Hello, world!", event.text)
    }

    @Test
    fun entityEnteredConstruction() {
        val event = WorldEvent.EntityEntered(
            roomId = "room-1",
            timestamp = now,
            entityId = "player-1",
            entityName = "Alice",
            entityType = "player",
            fromDirection = "north",
        )
        assertEquals("room-1", event.roomId)
        assertEquals(now, event.timestamp)
        assertEquals("player-1", event.entityId)
        assertEquals("Alice", event.entityName)
        assertEquals("player", event.entityType)
        assertEquals("north", event.fromDirection)
    }

    @Test
    fun typeDiscrimination() {
        val events: List<WorldEvent> = listOf(
            WorldEvent.RoomCreated("r", now, "Room", "Desc", "zone"),
            WorldEvent.Said("r", now, "p1", "Alice", "Hi"),
            WorldEvent.EntityEntered("r", now, "p1", "Alice", "player", "north"),
            WorldEvent.EntityLeft("r", now, "p1", "Alice", "south"),
        )

        var roomCreated = false
        var said = false
        var entered = false
        var left = false

        for (event in events) {
            when (event) {
                is WorldEvent.RoomCreated -> roomCreated = true
                is WorldEvent.Said -> said = true
                is WorldEvent.EntityEntered -> entered = true
                is WorldEvent.EntityLeft -> left = true
                else -> {} // other event types
            }
        }

        assertTrue(roomCreated, "RoomCreated should be matched")
        assertTrue(said, "Said should be matched")
        assertTrue(entered, "EntityEntered should be matched")
        assertTrue(left, "EntityLeft should be matched")
    }
}

File: clients/rn/__tests__/engine/world-event.test.ts
import type { WorldEvent, RoomCreated, Said, EntityEntered } from '../../src/engine/events/WorldEvent';

describe('WorldEvent', () => {
  it('creates a RoomCreated event', () => {
    const event: RoomCreated = {
      type: 'room_created', roomId: 'test', timestamp: Date.now(),
      name: 'Test Room', description: 'A test room.', zone: 'test-zone',
    };
    expect(event.type).toBe('room_created');
    expect(event.roomId).toBe('test');
    expect(event.name).toBe('Test Room');
  });

  it('creates a Said event', () => {
    const event: Said = {
      type: 'said', roomId: 'test', timestamp: Date.now(),
      entityId: 'p1', entityName: 'Alice', text: 'Hello!',
    };
    expect(event.type).toBe('said');
    expect(event.text).toBe('Hello!');
  });

  it('discriminates between event types', () => {
    const event: WorldEvent = {
      type: 'entity_entered', roomId: 'test', timestamp: Date.now(),
      entityId: 'p1', entityName: 'Alice', entityType: 'player', fromDirection: 'south',
    };
    if (event.type === 'entity_entered') {
      expect(event.entityName).toBe('Alice');
      expect(event.fromDirection).toBe('south');
    } else {
      fail('Expected entity_entered');
    }
  });

  it('all 15 event types are constructible', () => {
    const ts = Date.now();
    const events: WorldEvent[] = [
      { type: 'room_created', roomId: 'r', timestamp: ts, name: 'N', description: 'D', zone: 'Z' },
      { type: 'entity_entered', roomId: 'r', timestamp: ts, entityId: 'e', entityName: 'E', entityType: 'player', fromDirection: 'n' },
      { type: 'entity_left', roomId: 'r', timestamp: ts, entityId: 'e', entityName: 'E', direction: 'n' },
      { type: 'said', roomId: 'r', timestamp: ts, entityId: 'e', entityName: 'E', text: 'hi' },
      { type: 'object_taken', roomId: 'r', timestamp: ts, entityId: 'e', objectId: 'o', objectName: 'O' },
      { type: 'object_dropped', roomId: 'r', timestamp: ts, entityId: 'e', objectId: 'o', objectName: 'O', description: 'D', takeable: true },
      { type: 'object_used', roomId: 'r', timestamp: ts, entityId: 'e', objectId: 'o', objectName: 'O', target: null, result: null },
      { type: 'exit_opened', roomId: 'r', timestamp: ts, direction: 'n', targetRoom: 't', label: 'L' },
      { type: 'exit_closed', roomId: 'r', timestamp: ts, direction: 'n' },
      { type: 'description_changed', roomId: 'r', timestamp: ts, newDescription: 'New', reason: null },
      { type: 'hints_updated', roomId: 'r', timestamp: ts, hints: [] },
      { type: 'script_triggered', roomId: 'r', timestamp: ts, scriptName: 'S', trigger: 'T', context: {} },
      { type: 'object_added', roomId: 'r', timestamp: ts, objectId: 'o', objectName: 'O', description: 'D', takeable: false },
      { type: 'property_changed', roomId: 'r', timestamp: ts, key: 'k', oldValue: null, newValue: 'v' },
      { type: 'whispered', roomId: 'r', timestamp: ts, entityId: 'e1', entityName: 'E1', targetEntityId: 'e2', text: 'psst' },
    ];
    expect(events).toHaveLength(15);
    const types = new Set(events.map(e => e.type));
    expect(types.size).toBe(15);
  });
});
